gcsio/src/main/java/com/google/cloud/hadoop/gcsio/MetricsPublisher.java

```diff
@@ -0,0 +1,243 @@
+
+  /**
+   * Publishes metric for specified open-census measurement
+   *
+   * @param key represents filterable attribute
+   * @param value for the TagKey
+   * @param ml - Measurement to be recorded
+   * @param n - Value of the measurement recorded
+   */
+  void recordTaggedStat(TagKey key, String value, MeasureLong ml, Long n);
+
+  /**
+   * Publishes metric for specified open-census measurement
+   *
+   * @param keys - List of TagKeys attributes used for filtration
+   * @param values - corresponding values for list of TagKeys
+   * @param ml - Measurement to be recorded
+   * @param n - Value of the measurement recorded
+   */
+  void recordLong(TagKey[] keys, String[] values, MeasureLong ml, Long n);
+}
+
+/** Publishes open-census measurements to StackDriver */
+class StackDriverMetricsPublisher implements MetricsPublisher {
+
+  @VisibleForTesting static final Object monitor = new Object();
+
+  @GuardedBy("monitor")
+  private static boolean initialized = false;
+
+  private static final TagMetadata TAG_METADATA_NO_PROPAGATION =
+      TagMetadata.create(TagTtl.NO_PROPAGATION);
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  static final MeasureLong LATENCY_MS =
+      MeasureLong.create("gcsio/latency", "The latency in milliseconds ", "ms");
+
+  static final MeasureLong MESSAGE_LATENCY_MS =
+      MeasureLong.create(
+          "gcsio/message/latency", "The latency in milliseconds per gcs message loop", "ms");
+
+  static final MeasureLong REQUESTS =
+      MeasureLong.create(
+          "gcsio/requests", "The distribution of retry attempts for gcs calls", "By");
+
+  static final MeasureLong REQUEST_RETRIES =
+      MeasureLong.create("gcsio/retries", "The distribution of retry attempts for gcs calls", "By");
+
+  static final TagKey KEY_METHOD = TagKey.create("method");
+  static final TagKey KEY_STATUS = TagKey.create("status");
+  static final TagKey KEY_ERROR = TagKey.create("error");
+  static final TagKey KEY_THREAD = TagKey.create("thread");
+  static final TagKey KEY_PROTOCOL = TagKey.create("protocol");
+
+  static final List<TagKey> TAG_KEYS =
+      Collections.unmodifiableList(
+          Arrays.asList(KEY_METHOD, KEY_STATUS, KEY_ERROR, KEY_THREAD, KEY_PROTOCOL));
+
+  private static final Tagger tagger = Tags.getTagger();
+  private static final StatsRecorder statsRecorder = Stats.getStatsRecorder();
+
+  public static MetricsPublisher create(String projectId, Credential credential) {
+    CredentialAdapter credentialAdapter = new CredentialAdapter(credential);
+    return create(projectId, credentialAdapter);
+  }
+
+  public static MetricsPublisher create(String projectId, Credentials credentials) {
+    try {
+      registerAllViews();
+      setupStackDriverExporter(projectId, credentials);
+    } catch (Exception e) {
+      logger.atWarning().withCause(e).log("Exception while registering metrics publisher");
+      return new NoOpMetricsPublisher();
```

Comment:
Of you are throwing exception then you do not need to log, just attach cause to it.

Reply:
We are not throwing an exception here, We rely on No-Op